src/hooks.js

```diff
@@ -41,32 +41,57 @@ export function useArrows({ onUp, onLeft, onDown, onRight } = {}) {
   }, [onUp, onLeft, onDown, onRight])
 }
 
+function stepsReducer(state, { type, value, steps }) {
+  const { step } = state
+
+  let newStep = null
+
+  if (type === 'set') {
+    newStep = value
+  }
+  if (type === 'next' && step < steps - 1) {
+    newStep = step + 1
+  }
+  if (type === 'prev' && step > 0) {
+    newStep = step - 1
+  }
+
+  if (newStep !== null && step !== newStep) {
+    return {
+      step: newStep,
+      direction: newStep > step ? 1 : -1,
+    }
+  }
+
+  return state
+}
+
 // Simple hook to manage a given number of steps.
 export function useSteps(steps) {
-  const [step, _setStep] = useState(0)
-  const [direction, setDirection] = useState(0)
+  const [{ step, direction }, updateStep] = useReducer(stepsReducer, {
+    step: 0,
+    direction: 0,
+  })
+
+  // If the number of steps change, we reset the current step
+  useEffect(() => {
+    updateStep({ type: 'set', value: 0, steps })
+  }, [steps])
 
   const setStep = useCallback(
-    newStep => {
-      if (step !== newStep) {
-        setDirection(newStep > step ? 1 : -1)
-        _setStep(newStep)
-      }
+    value => {
+      updateStep({ type: 'set', value, steps })
     },
-    [setDirection, _setStep, step]
+    [steps]
```

Comment:
Do we not need to include the reducer's `updateStep` because it'll never change?

Reply:
Exactly, `setSomething` from `useState()` and `dispatch` from `useReducer()` are both guaranteed to never update [1], and don’t need to be added to the dependencies list.

[1] Search for “guarantees” in these two sections:

https://reactjs.org/docs/hooks-reference.html#usestate
https://reactjs.org/docs/hooks-reference.html#usereducer